### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by rdehuyss** [gadgetchnnel/lovelace-home-feed-card#16]
Just wanted to chip in on the performance issue: I can also confirm it. With me, chrome is taking 77% and homeassistant 50% when using this card. I'm using the dev version installed with HACS.

What I notice is a whole lot of calls to `http://127.0.0.1:8123/api/history/period?filter_entity_id=` each 10 seconds although I only have two entities.

```
      - type: 'custom:home-feed-card'
        title: Home Feed
        show_empty: true
        more_info_on_tap: true
        entities:
          - entity: binary_sensor.voordeur
            exclude_states: 
              - "off" 
              - "unknown"
          - entity: binary_sensor.bureau_schuifraam_1
            exclude_states: 
              - "off" 
              - "unknown"
```

![Screenshot from 2019-09-05 09-39-25](https://user-images.githubusercontent.com/567842/64321577-19343380-cfc1-11e9-9103-29d7e1e44ebd.png)

Removing the above configuration from my lovelace and cpu usage is back normal.

**Edit 1**
Some more info: each time HomeAssistant sends an Update through set hass, the stack is as follows:
```
- set hass(hass)
- refreshEntityHistory
- getLiveEntityHistory *which makes http calls even if entity_ids is empty*
```

I cannot completely follow the logic/code of your card since HomeAssistant provides you with updates of everything by calling set hass(hass) => hass contains the latest state and instead of doing http calls each time, can't you store the changes yourself in local_storage? This would eliminate all the http calls and thus solve the performance issue I think.
Anyway: set hass(hass) can happen several times per second and since the history api is rather slow, I at least would throttle the calls to history/period.

Cheers!

**Comment by gadgetchnnel** [comment]
> - getLiveEntityHistory *which makes http calls even if entity_ids is empty*
I have just fixed that in 0.2.11b4, it now only calls the history API if there are entities with **include_history** set.

> I cannot completely follow the logic/code of your card since HomeAssistant provides you with updates of everything by calling set hass(hass) => hass contains the latest state and instead of doing http calls each time, can't you store the changes yourself in local_storage?

Doing it entirely that way would rely on any device you access the UI from being constantly online. If you only occasionally access it on your phone, for example, you will potentially miss events which happened since you last accessed it. Using the history API avoids this.

I'm currently investigating solutions to avoid too many calls to the history API. The most likely is that entities with history will display the latest state "live" from the **hass** object, and refresh the history less frequently. This will possibly be based on when the state was last changed (e.g. only if it is at least 1 minute since the last updated time of that entity).

**Comment by rdehuyss** [comment]
Ok, makes sense to use the history api then. And what about doing the
history call once for each entity and then use the hass object to monitor
for state changes?

**Comment by rdehuyss** [comment]
I did some performance analysis using chrome and I think I found the root cause: there is an endless loop in your code.

![Screenshot from 2019-09-06 09-58-46](https://user-images.githubusercontent.com/567842/64411177-f5d6ba80-d08c-11e9-99de-74987ba7bf88.png)

So, what happens: 
- you call buildIfReady()
- this calls refreshNotifications()
- and in refreshNotifications() you again call buildIfReady() thus creating an endless loop...

And although you have a break clause in your code (see below), this is only working if people configure calendar events which I think you have configured, but the other people not (me :wink: )
```
        if((!this.loadedNotifications || !notificationsLastUpdate) && this.moment){
    		this.refreshNotifications().then(() => {});
    	}
```

This makes my chrome browser go :banana: . I don't know yet why the backend is having troubles though.

**Comment by rdehuyss** [comment]
I can confirm that if I comment line 773 all performance issues are solved, both frontend and backend.

Before:
![before](https://user-images.githubusercontent.com/567842/64417563-9df28080-d099-11e9-93a7-dedd06ba60ef.png)


After
![after](https://user-images.githubusercontent.com/567842/64417570-a3e86180-d099-11e9-8d4a-19669b0ebb79.png)

I hope this helps you solve the issue!

Cheers

**Comment by gadgetchnnel** [comment]
@rdehuyss Thanks for the information. That was a great help. It seems there was a typo there.
notificationsLastUpdate should have actually been using **home-feed-card-notificationsLastUpdate** from local storage but it was using **home-feed-card-events** instead which, as you point out, means it will cause it to keep trying if you don't have calendars configured.

I have deployed a fix for this to 0.2.11b5

**Comment by rdehuyss** [comment]
Wonderful! Happy I could help.

I did one other small suggestion (see PR) and if you would be willing to accept the PR, I can confirm if the fix is working. I now have a copy of your card running my changes.

**Comment by gadgetchnnel** [comment]
@rdehuyss I've merged your PR and included this in the latest beta release, 0.2.11b6
